checknoglobals/check_no_globals.go

```diff
@@ -96,6 +96,31 @@ func looksLikeError(i *ast.Ident) bool {
 	return strings.HasPrefix(i.Name, prefix)
 }
 
+// hasEmbedComment returns true if the AST identifier has
+// a '//go:embed ' comment, or false otherwise.
+func hasEmbedComment(i *ast.Ident) bool {
+	if i.Obj == nil {
+		return false
+	}
+
+	spec, ok := i.Obj.Decl.(*ast.ValueSpec)
+	if !ok {
+		return false
+	}
+
+	if spec.Doc == nil {
+		return false
+	}
+
+	for _, c := range spec.Doc.List {
```

Comment:
❓ Is the `.List` each line of comment above the function? If that's the case I think we just need to get the last line.

Reply:
Ah so I thought that the embed directive comment had to immediately proceed the declaration, but it turns out there are allowed to be empty comment lines between the two, so checking all comment lines is probably necessary.

> The directive must immediately precede a line containing the declaration of a single variable. Only blank lines and ‘//’ line comments are permitted between the directive and the declaration. 

Ref: https://tip.golang.org/pkg/embed/#hdr-Directives

Reply:
Yes I was thinking the same thing. But the following is valid and works (I tested it):

```go
	//go:embed swagger/*
	// a comment
	//
	//
	static embed.FS //nolint: gochecknoglobals // no other possibility
```

Even adding an empty line is valid:

```go
	//go:embed swagger/*

	static embed.FS //nolint: gochecknoglobals // no other possibility
```

With this empty line, my solution does not work anymore, because in the `isAllowed(v ast.Node) ` the Node has no comment. For testing purposes I printed all Nodes with [spew](https://github.com/davecgh/go-spew):

```go
spew.Dump(v)
```

And the directive comment never showed up in the  `isAllowed(v ast.Node) ` method. So with the current method it is not possible to make an exception for the above. Is the parsing done with the [parse.ParseComments](https://golang.org/pkg/go/parser/#Mode) mode?

Maybe we can live with that? I think most devs do not put an empty line between the directive and the actual variable declaration.

Reply:
You should be able to get the case with an empty line working by getting the comment map per node like this:

```go
ast.NewCommentMap(fset, node, file.Comments)
```

Where `fset` is a `*token.FileSet` that's public from each pass with `pass.Fset` and `file` is a `*ast.File` which is the current file when iterating over `pass.Files`.

Reply:
Is it leaving an empty line between the variable and the comments considered a good practice? Why accept that in the first place?

Reply:
I would say no and don't see any reason to allow it. But if it's allowed and working specifically for `//go:embed` directives according to the go source I guess it would be an alternative to allow it here as well.

Reply:
I would leave it like this and add it later if someone complains.

Reply:
Thanks to @bombsimon 's suggestion I have a solution, that also works with empty lines. For that I had to change the signature from:

```go
func isAllowed(v ast.Node) bool 
```

to

```go
func isAllowed(v ast.Node, fset *token.FileSet, file *ast.File) bool 
```

If you want I can update my MR.

Reply:
I think you should add this. If it's allowed in the embed package which is what this PR is about to address, it should be allowed here as well. `hasEmbedComment` so it shouldn't be an issue changing the signature. I don't think it's good practice but I also don't think this linter should address bad practices regarding comments.

Reply:
I think it is worth adding this too. +1

Reply:
I tried with @bombsimon solution. The problem I have is, that with `ast.NewCommentMap(fset, node, file.Comments)` I get all comments, not just for the current node.

I also checked go's source code. The compiler does not use ast to find embeded files. It is using a custom (internal) reader.

So I suggest using the actual pull request (which does not work when there is an empty line after the `embed` directive).

Reply:
> I get all comments, not just for the current node.

Do you mean you get all comments for the entire file, or just all comments between the current node and the previous node?

Reply:
I won't have time to look at this today but the map should contain all comments (you pass `file.Comments`) but you should be able to get a subset by accessing the map with your current `ast.Node`.

This looks a bit wrong, it will iterate over **all** the file comments like you describe:

```go
cm := ast.NewCommentMap(fset, i, file.Comments)
for _, c := range cm.Comments() ...
```

It should be something like this where you first ensure the comment is related to your node:

```go
cm := ast.NewCommentMap(fset, i, file.Comments)

commentsRelatedToNode, ok := cm[i] // For the passed nodoe/ident
if !ok {
    return false
}

for _, c := range commentsRelatedToNode ... // Only parse comments related to this node
```

Can't remember exactly from the top of my head but you need to access your specific node to make proper use of the comment map. If you haven't figured this out tomorrow I'll try to have a look later this week as long as you've pushed your latest changes.

Reply:
@bombsimon I tried the following:

```go
commentsRelatedToNode, ok := cm[i] // For the passed nodoe/ident
```

But it did not work. I verify this afternoon again. I let you know.

Reply:
OK, I verified again. For the following test file:

```go
package code

import (
	"embed"
)

var blo = 1 // want "myVar is a global variable"

// Testing embedded data, all allowed.
var (
	//go:embed dummy/*
	asserts embed.FS

	//go:embed public/*

	public embed.FS

	//go:embed dummy/file.txt
	data []byte
)
```

With the following code:

```go
func hasEmbedComment(n ast.Node, fset *token.FileSet, file *ast.File) bool {
	cm := ast.NewCommentMap(fset, n, file.Comments)

	commentsRelatedToNode, ok := cm[n]
	if !ok {
		return false
	}

	for i := range commentsRelatedToNode {
		if commentsRelatedToNode[i].Text() != "" {
			fmt.Println(commentsRelatedToNode[i].Text())
		}
		for _, c := range commentsRelatedToNode[i].List {
			fmt.Println(c.Text)
		}
	}
	fmt.Println("-------------")

	return false
}
```

I get the following output:

```console
want "myVar is a global variable"                                                                                                                                                                                                                                                                                                                      

// want "myVar is a global variable"
Testing embedded data, all allowed.

// Testing embedded data, all allowed.
//go:embed dummy/*
//go:embed public/*
//go:embed dummy/file.txt
-------------
want "myVar is a global variable"

// want "myVar is a global variable"
Testing embedded data, all allowed.

// Testing embedded data, all allowed.
//go:embed dummy/*
//go:embed public/*
//go:embed dummy/file.txt
-------------
want "myVar is a global variable"

// want "myVar is a global variable"
Testing embedded data, all allowed.

// Testing embedded data, all allowed.
//go:embed dummy/*
//go:embed public/*
//go:embed dummy/file.txt
-------------
want "myVar is a global variable"

// want "myVar is a global variable"
Testing embedded data, all allowed.

// Testing embedded data, all allowed.
//go:embed dummy/*
//go:embed public/*
//go:embed dummy/file.txt
-------------
```

To me it looks like I always get all comments. Not sure what I am doing wrong.